lib/test.js

```diff
@@ -550,14 +551,28 @@ Test.prototype._runChild = function runChild (child, name, extra, cb) {
 
   // still need try/catch for synchronous errors
   try {
+    child._mustDeferEnd = true;
     var cbRet = cb(child)
     if (cbRet && typeof cbRet.then === 'function') {
       // promise
       cbRet.then(function () {
-        child.end(IMPLICIT)
-      }, child.threw)
+        child._mustDeferEnd = false;
+        if (!child._tryResumeEnd()) {
+          child.end(IMPLICIT)
+        }
+      }, function (err) {
+        child._mustDeferEnd = false;
+        if (!child._tryResumeEnd()) {
```

Comment:
Feel free to dismiss this as a personal style preference, but I re-factored this chunk when exploring locally to look like this:
```js
  // still need try/catch for synchronous errors
  try {
    child._mustDeferEnd = true;
    var cbRet = cb(child)
    if (cbRet && typeof cbRet.then === 'function') {
      // promise
      cbRet.then(callbackResolved, callbackThrew)
    } else {
      callbackFinished()
    }
  } catch (er) {
    callbackThrew(er)
  }
  self._level = self

  function callbackFinished() {
    child._mustDeferEnd = false
    return child._tryResumeEnd()
  }

  function callbackThrew(err) {
    callbackFinished()
    child.threw(err);
  }

  function callbackResolved() {
    if (!callbackFinished()) {
      child.end(IMPLICIT)
    }
  }
```